File: CyFile-Android/app/src/main/java/at/tugraz/tc/cyfile/crypto/impl/KeyVaultServiceImpl.java
package at.tugraz.tc.cyfile.crypto.impl;

import android.security.keystore.KeyGenParameterSpec;
import android.security.keystore.KeyProperties;

import java.io.IOException;
import java.security.InvalidAlgorithmParameterException;
import java.security.Key;
import java.security.KeyStore;
import java.security.KeyStoreException;
import java.security.NoSuchAlgorithmException;
import java.security.NoSuchProviderException;
import java.security.UnrecoverableKeyException;
import java.security.cert.CertificateException;

import javax.crypto.KeyGenerator;

import at.tugraz.tc.cyfile.crypto.KeyVaultService;
import at.tugraz.tc.cyfile.crypto.exceptions.InvalidPassPhraseException;
import at.tugraz.tc.cyfile.crypto.exceptions.KeyVaultAlreadyInitializedException;
import at.tugraz.tc.cyfile.crypto.exceptions.KeyVaultLockedException;
import at.tugraz.tc.cyfile.crypto.exceptions.KeyVaultNotInitializedException;
import at.tugraz.tc.cyfile.crypto.exceptions.KeyVaultServiceException;

public class KeyVaultServiceImpl implements KeyVaultService {

    private Key secretKey;
    private static final String KEY_ALIAS = "cyfile-encryption-key";
    private static final String KEYSTORE_PROVIDER = "AndroidKeyStore";
    private State internalState = State.INIT;

    private KeyGenerator generator;
    private KeyStore keyStore;

    private enum State {
        INIT,
        UNLOCKED,
        LOCKED
    }

    public KeyVaultServiceImpl(KeyGenerator keyGenerator, KeyStore keyStore) {
        this.secretKey = null;
        this.keyStore = keyStore;
        this.generator = keyGenerator;
    }

    public KeyVaultServiceImpl() {
        this.secretKey = null;
        try {
            this.keyStore = KeyStore.getInstance(KEYSTORE_PROVIDER);
        } catch (KeyStoreException e) {
            throw new KeyVaultServiceException("KeyStore cannot be initialized");
        }
        try {
            generator = KeyGenerator
                    .getInstance(KeyProperties.KEY_ALGORITHM_AES, KEYSTORE_PROVIDER);
        } catch (NoSuchAlgorithmException | NoSuchProviderException e) {
            throw new KeyVaultServiceException("Configuration failure");
        }

        final KeyGenParameterSpec keyGenParameterSpec = new KeyGenParameterSpec.Builder(KEY_ALIAS,
                KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT)
                .setBlockModes(KeyProperties.BLOCK_MODE_CBC)
                .setKeySize(256)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7)
                .build();

        try {
            generator.init(keyGenParameterSpec);
        } catch (InvalidAlgorithmParameterException e) {
            throw new KeyVaultServiceException("Wrong parameter configuration");
        }
    }

    @Override
    public void init(String passphrase) throws KeyVaultAlreadyInitializedException {
        if (internalState != State.INIT) {
            throw new KeyVaultAlreadyInitializedException();
        }


        secretKey = generator.generateKey();
        internalState = State.LOCKED;
    }

    @Override
    public void unlockVault(String passphrase) throws KeyVaultNotInitializedException, InvalidPassPhraseException {
        if (internalState == State.INIT) {
            throw new KeyVaultNotInitializedException("Key vault must be initialized first.");
        }

        // already unlocked
        if (internalState == State.UNLOCKED) {
            return;
        }

        try {
            this.keyStore.load(null);
            if (this.keyStore.containsAlias(KEY_ALIAS)) {
                // load key from keystore
                secretKey = this.keyStore.getKey(KEY_ALIAS, passphrase.toCharArray());
            } else {
                throw new KeyVaultServiceException("Configuration is invalid - contact programmer!");

            }
        } catch (IOException | NoSuchAlgorithmException | CertificateException | KeyStoreException | UnrecoverableKeyException e) {
            throw new InvalidPassPhraseException("Passphrase is invalid");
        }

        internalState = State.UNLOCKED;

    }

    @Override
    public void lockVault() throws KeyVaultNotInitializedException {
        if (internalState == State.LOCKED) {
            return;
        }

        if (internalState == State.INIT) {
            throw new KeyVaultNotInitializedException("Must be initialized first");
        }

        internalState = State.LOCKED;
    }

    @Override
    public Key getEncryptionKey() throws KeyVaultLockedException, KeyVaultNotInitializedException {
        if (internalState == State.INIT) {
            throw new KeyVaultNotInitializedException("Not initialized");
        }

        if (internalState == State.LOCKED) {
            throw new KeyVaultLockedException("Key vault is locked");
        }

        return secretKey;
    }
}

File: CyFile-Android/app/src/main/kotlin/at/tugraz/tc/cyfile/crypto/impl/KeyVaultServiceImpl.kt
package at.tugraz.tc.cyfile.crypto.impl

import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import at.tugraz.tc.cyfile.crypto.KeyVaultService
import at.tugraz.tc.cyfile.crypto.exceptions.*
import at.tugraz.tc.cyfile.logging.CyFileLogger
import java.io.IOException
import java.security.*
import java.security.cert.CertificateException
import javax.crypto.KeyGenerator

class KeyVaultServiceImpl : KeyVaultService {


    private var secretKey: Key? = null
    private var internalState = State.INIT

    private var logger: CyFileLogger
    private var generator: KeyGenerator
    private var keyStore: KeyStore

    private enum class State {
        INIT,
        UNLOCKED,
        LOCKED
    }

    constructor(log: CyFileLogger) {
        this.secretKey = null
        this.logger = log
        try {
            this.keyStore = KeyStore.getInstance(KEYSTORE_PROVIDER)
        } catch (e: KeyStoreException) {
            throw KeyVaultServiceException("KeyStore cannot be initialized")
        }
        try {
            generator = KeyGenerator
                    .getInstance(KeyProperties.KEY_ALGORITHM_AES, KEYSTORE_PROVIDER)
        } catch (e: NoSuchAlgorithmException) {
            throw KeyVaultServiceException("Configuration failure")
        } catch (e: NoSuchProviderException) {
            throw KeyVaultServiceException("Configuration failure")
        }
        val keyGenParameterSpec = KeyGenParameterSpec.Builder(KEY_ALIAS,
                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT)
                .setBlockModes(KeyProperties.BLOCK_MODE_CBC)
                .setKeySize(256)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7)
                .build()
        try {
            generator.init(keyGenParameterSpec)
        } catch (e: InvalidAlgorithmParameterException) {
            throw KeyVaultServiceException("Wrong parameter configuration")
        }
    }

    constructor(keyGenerator: KeyGenerator, keyStore: KeyStore, logger: CyFileLogger) {
        this.logger = logger
        this.generator = keyGenerator
        this.keyStore = keyStore
    }

    @Throws(KeyVaultAlreadyInitializedException::class)
    override fun init(passphrase: String) {
        if (internalState != State.INIT) {
            throw KeyVaultAlreadyInitializedException()
        }


        try {
            this.keyStore.load(null)
            if (!keyStore.containsAlias(KEY_ALIAS)) {
                this.logger.d("KeyVaultService", "Generating key")
                this.secretKey = generator.generateKey()
            }
            internalState = State.LOCKED
        } catch (e: KeyStoreException) {
            e.printStackTrace()

            // if error occurs than just leave it in init state
            internalState = State.INIT

        } catch (e: CertificateException) {
            e.printStackTrace()
            internalState = State.INIT
        } catch (e: NoSuchAlgorithmException) {
            e.printStackTrace()
            internalState = State.INIT
        } catch (e: IOException) {
            e.printStackTrace()
            internalState = State.INIT
        }

    }

    @Throws(KeyVaultNotInitializedException::class, InvalidPassPhraseException::class)
    override fun unlockVault(passphrase: String) {
        if (internalState == State.INIT) {
            throw KeyVaultNotInitializedException("Key vault must be initialized first.")
        }

        // already unlocked
        if (internalState == State.UNLOCKED) {
            return
        }

        try {
            this.keyStore.load(null)
            if (this.keyStore.containsAlias(KEY_ALIAS)) {
                // load key from keystore
                secretKey = this.keyStore.getKey(KEY_ALIAS, passphrase.toCharArray())
            } else {
                throw KeyVaultServiceException("Configuration is invalid - contact programmer!")
            }
        } catch (e: IOException) {
            throw InvalidPassPhraseException("Passphrase is invalid", e)
        } catch (e: NoSuchAlgorithmException) {
            throw InvalidPassPhraseException("Passphrase is invalid", e)
        } catch (e: CertificateException) {
            throw InvalidPassPhraseException("Passphrase is invalid", e)
        } catch (e: KeyStoreException) {
            throw InvalidPassPhraseException("Passphrase is invalid", e)
        } catch (e: UnrecoverableKeyException) {
            throw InvalidPassPhraseException("Passphrase is invalid", e)
        }

        logger.d("KeyVaultService", "KVS unlocked")
        internalState = State.UNLOCKED

    }

    @Throws(KeyVaultNotInitializedException::class)
    override fun lockVault() {
        if (internalState == State.LOCKED) {
            return
        }

        if (internalState == State.INIT) {
            throw KeyVaultNotInitializedException("Must be initialized first")
        }

        internalState = State.LOCKED
    }

    override fun getEncryptionKey(): Key? {
        if (internalState == State.INIT) {
            throw KeyVaultNotInitializedException("Not initialized")
        }

        if (internalState == State.LOCKED) {
            throw KeyVaultLockedException("Key vault is locked")
        }

        return secretKey
    }


    companion object {
        private const val KEY_ALIAS = "cyfile-encryption-key"
        private const val KEYSTORE_PROVIDER = "AndroidKeyStore"
    }

    init {

    }
}
